core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushJoinIntoTableScan.java

```diff
@@ -0,0 +1,270 @@
+    private static final Capture<TableScanNode> RIGHT_TABLE_SCAN = newCapture();
+
+    private static final Pattern<JoinNode> PATTERN =
+            Patterns.join()
+                    .with(Patterns.Join.left().matching(tableScan().capturedAs(LEFT_TABLE_SCAN)))
+                    .with(Patterns.Join.right().matching(tableScan().capturedAs(RIGHT_TABLE_SCAN)));
+
+    private final Metadata metadata;
+
+    public PushJoinIntoTableScan(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(JoinNode joinNode, Captures captures, Context context)
+    {
+        if (getJoinPushdownMode(context.getSession()) == JoinPushdownMode.OFF) {
+            return Result.empty();
+        }
+
+        if (joinNode.isCrossJoin()) {
+            return Result.empty();
+        }
+
+        TableScanNode left = captures.get(LEFT_TABLE_SCAN);
+        TableScanNode right = captures.get(RIGHT_TABLE_SCAN);
+
+        Expression effectiveFilter = getEffectiveFilter(joinNode);
+        FilterSplitResult filterSplitResult = splitFilter(effectiveFilter, joinNode.getLeftOutputSymbols(), joinNode.getRightOutputSymbols(), context);
+
+        if (!filterSplitResult.getRemainingFilter().equals(BooleanLiteral.TRUE_LITERAL)) {
+            // TODO add extra filter node above join
+            return Result.empty();
+        }
+
+        Map<String, ColumnHandle> leftAssignments = left.getAssignments()
+                .entrySet().stream()
+                .collect(toImmutableMap(entry -> entry.getKey().getName(), Map.Entry::getValue));
+
+        Map<String, ColumnHandle> rightAssignments = right.getAssignments()
+                .entrySet().stream()
+                .collect(toImmutableMap(entry -> entry.getKey().getName(), Map.Entry::getValue));
+
+        Optional<JoinApplicationResult<TableHandle>> joinApplicationResult = metadata.applyJoin(
+                context.getSession(),
+                getJoinType(joinNode),
+                left.getTable(),
+                right.getTable(),
+                filterSplitResult.getPushableConditions(),
+                // TODO we could send only subset of assignments here, those which are needed to resolve filterSplitResult.getPushableConditions
+                leftAssignments,
+                rightAssignments);
+
+        if (joinApplicationResult.isEmpty()) {
+            return Result.empty();
+        }
+
+        TableHandle handle = joinApplicationResult.get().getTableHandle();
+
+        Map<ColumnHandle, ColumnHandle> columnHandlesMapping = joinApplicationResult.get().getColumnHandles();
+
+        TupleDomain<ColumnHandle> newEnforcedConstraints = TupleDomain.fromColumnDomains(
+                transformToNewAssignments(left.getEnforcedConstraint(), columnHandlesMapping).getColumnDomains().flatMap(
```

Comment:
the `getColumnDomains` being called here has `// Available for Jackson serialization only!` comment
i guess it should be marked deprecated as in https://github.com/trinodb/trino/blob/02b59cd274f5314ef3c65ef0064b0371da449537/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcColumnHandle.java#L53-L55
(out of scope for this PR)

and definitely not used here